### PR TITLE
Implement ASIFT in C++

### PR DESCRIPTION
**Pull request by Yosshi999**
ASIFT in C++.

The reason of the redundant test script is that it is for a future testdata generated by Python implementation (asift.py). Filestorage in python's opencv is supporting only cv::Mat. So this test uses cv::Mat for storing/reading generated data.

I referred:
- Python implementation in https://github.com/opencv/opencv/blob/3.4/samples/python/asift.py
- C++ implementation in https://github.com/kukuruza/Erie

Actually my implementation is different from Erie, but I used some idea from that.
I have no idea how to specify its contribution so please give me some advice.

This PR is related to Issue #4399 .
Test data is in https://github.com/opencv/opencv_extra/pull/775

## performance comparison
I compared the extraction time for the 1st image (aero1.jpg) described in asift.py.
For C++ backend, I measured the execution time by running `cv2.AffineFeature_create` and `detectAndCompute` in python.
I used `time.perf_counter()` for measuring the time.

The unit is `sec`.
| feature option | affine_detect() in asift.py | C++ backend | speed up 
| -- | -- | -- | --
| brisk | 0.795482 | 0.484333 | 1.64
| sift | 2.134979 |1.843431 | 1.16
| orb | 0.436688 | 0.327703 | 1.33

## degree of similarity between Python version and this
I was not able to exactly reproduce the keypoints which asift.py generate. This may be because of floating-point precision (Python uses double) and some numpy functions.
However, they are almost similar. See https://github.com/terfendail/opencv/pull/9#issuecomment-659134122 for comparison.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by tompollok** [comment]
Great work. Will you also add a cpp sample for asift usage?

**Comment by sturkmen72** [comment]
thanks for your work. i tested the sample code and want to share my  remarks

1. when i used following images the program exits without any message and doing nothing
http://demo.ipol.im/demo/my_affine_sift/tmp/FA9F327A9B3DCB99D1332D9D266AB58E/input_0.png
http://demo.ipol.im/demo/my_affine_sift/tmp/FA9F327A9B3DCB99D1332D9D266AB58E/input_1.png

2. is it possible to draw the polylines around the object
affine find_obj

File: modules/features2d/src/affine_feature.cpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    void affineSkew(float tilt, float phi,
+            Mat& warpedImage, Mat& warpedMask, Matx23f& pose) const
+    {
+        int h = image.size().height;
+        int w = image.size().width;
+        Mat rotImage;
+
+        Mat mask0;
+        if( mask.empty() )
+            mask0 = Mat::ones(h, w, CV_8UC1)*255;
```
Comment (sturkmen72) [inline modules/features2d/src/affine_feature.cpp:224]: `mask0 = Mat(h, w, CV_8UC1, Scalar(255));` seems better to me

File: samples/cpp/asift.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    img1.copyTo(Mat(vis, Rect(0, 0, w1, h1)));
+    img2.copyTo(Mat(vis, Rect(w1, 0, w2, h2)));
+    cvtColor(vis, vis, COLOR_GRAY2BGR);
+
+    vector<Point2f> corners{{0,0}, {(float)w1,0}, {(float)w1,(float)h1}, {0,(float)h1}};
+    vector<Point2i> icorners;
+    perspectiveTransform(corners, corners, H);
+    transform(corners, corners, Matx23f(1,0,w1,0,1,0));
+    Mat(corners).convertTo(icorners, CV_32S);
+    polylines(vis, {icorners}, true, Scalar(255,255,255));
```
Comment (sturkmen72) [inline samples/cpp/asift.cpp:152]: do we need {icorners}
`polylines(vis, icorners, true, Scalar(255,255,255)); ` worked for me

**Comment by Yosshi999** [comment]
> 1. when i used following images the program exits without any message and doing nothing

Thank you for testing. I've found a new bug in affine_feature.cpp.

> 2. is it possible to draw the polylines around the object

I will add an option to suppress the number of lines.

**Comment by Yosshi999** [comment]
I can't understand why this build bot returns failure. Is it because the testdata in my PR of opencv_extra is too large?

**Comment by sturkmen72** [comment]
there is a problem with document builder but seems unrelated with your commits. i think devs will solve it soon.

**Comment by sturkmen72** [comment]
oh sorry i looked wrong PR. yes because the testdata in your PR of opencv_extra is too large

**Comment by Yosshi999** [comment]
Oh I had a misunderstanding. This size limit is not per-file!

**Comment by mshabunin** [comment]
Is it possible to reduce test data size? Is it possible to reuse existing features2d tests, e.g. these: https://github.com/opencv/opencv/blob/0524fcb7b7beb4e4d050b65ee0f21f83db6084f2/modules/features2d/test/test_detectors_regression.cpp ?

**Comment by Yosshi999** [comment]
By default, the testdata generated by ASIFT is about 30 times larger than normal SIFT's one because ASIFT sums up 43 variants by extracting SIFT descriptors from various affine-warped images.
I will reduce the parameters of ASIFT.

File: modules/features2d/src/affine_feature.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
```
Comment (alalek) [inline modules/features2d/src/affine_feature.cpp:2]: Please add OpenCV short license header above with note that:

"This file is based on code issued with the following license"

File: modules/features2d/src/affine_feature.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
+/*********************************************************************
```
Comment (Yosshi999) [inline modules/features2d/src/affine_feature.cpp:1]:
```suggestion
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
// This file is based on code issued with the following license.
/*********************************************************************
```

Comment (Yosshi999) [reply modules/features2d/src/affine_feature.cpp:1]: @alalek Is this license header right?

Comment (alalek) [reply modules/features2d/src/affine_feature.cpp:1]: Yes (please add empty line before note)

**Pull request merged**